hugr-core/src/import.rs

```diff
@@ -1263,6 +1291,116 @@ impl<'a> Context<'a> {
+                    } else {
+                        let runtime_type =
+                            args.first().ok_or(model::ModelError::TypeError(term_id))?;
+                        let runtime_type = self.import_type(*runtime_type)?;
+
+                        let extensions =
+                            args.get(2).ok_or(model::ModelError::TypeError(term_id))?;
+                        let extensions = self.import_extension_set(*extensions)?;
+
+                        let value: serde_json::Value = serde_json::from_str(json)
+                            .map_err(|_| model::ModelError::TypeError(term_id))?;
+                        let custom_const = CustomSerialized::new(runtime_type, value, extensions);
+                        let opaque_value = OpaqueValue::new(custom_const);
+                        return Ok(Value::Extension { e: opaque_value });
+                    }
+                }
+
+                Err(error_unsupported!("constant value that is not JSON data"))
+                // TODO: This should ultimately include the following cases:
+                // - function definitions
+                // - custom constructors for values
+            }
+
+            model::Term::StaticType
+            | model::Term::Constraint
+            | model::Term::Const { .. }
+            | model::Term::List { .. }
+            | model::Term::ListType { .. }
+            | model::Term::Str(_)
+            | model::Term::StrType
+            | model::Term::Nat(_)
+            | model::Term::NatType
+            | model::Term::ExtSet { .. }
+            | model::Term::ExtSetType
+            | model::Term::Adt { .. }
+            | model::Term::FuncType { .. }
+            | model::Term::Control { .. }
+            | model::Term::ControlType
+            | model::Term::Type
+            | model::Term::NonLinearConstraint { .. } => {
+                Err(model::ModelError::TypeError(term_id).into())
+            }
+
+            model::Term::ConstFunc { .. } => Err(error_unsupported!("constant function value")),
+
+            model::Term::ConstAdt { tag, values } => {
+                let model::Term::Adt { variants } = self.get_term(type_id)? else {
+                    return Err(model::ModelError::TypeError(term_id).into());
+                };
+
+                let values = self.import_closed_list(*values)?;
+                let variants = self.import_closed_list(*variants)?;
+
+                let variant = variants
+                    .get(*tag as usize)
+                    .ok_or(model::ModelError::TypeError(term_id))?;
+                let variant = self.import_closed_list(*variant)?;
+
+                let items = values
+                    .iter()
+                    .zip(variant.iter())
+                    .map(|(value, typ)| self.import_value(*value, *typ))
+                    .collect::<Result<Vec<_>, _>>()?;
+
+                let typ = {
+                    // TODO: Import as a `SumType` directly and avoid the copy.
```

Comment:
TODO now or later?

Reply:
Later. This only affects performance and doesn't need any breaking changes to do right.